File: python/a19.py
# A19 Knapsack 1
# submission(PyPy3): https://atcoder.jp/contests/tessoku-book/submissions/38503186

n, w = map(int, input().split())
wv = [tuple(map(int, input().split())) for i in range(n)]

dp = [[0] * (w + 1) for i in range(n + 1)]

for i in range(n):
    for j in range(w + 1):
        if j + wv[i][0] <= w:
            dp[i + 1][j + wv[i][0]] = max(dp[i + 1][j + wv[i][0]], dp[i][j] + wv[i][1])
        dp[i + 1][j] = max(dp[i + 1][j], dp[i][j])

print(dp[n][w])

File: rust/tessoku-book/src/bin/a19.rs
// A19 Knapsack 1
// submission: https://atcoder.jp/contests/tessoku-book/submissions/38504604

use proconio::input;

#[macro_export]
macro_rules! chmax {
    ( $a:expr, $b:expr ) => {
        {
            if $a < $b {
                $a = $b;
                true
            } else {
                false
            }
        }
    };
}

fn main() {
    input!{
        n: usize,
        w: usize,
        wv:[(usize, usize); n],
    }
    let mut dp = vec![vec![0; w + 1]; n + 1];
    for i in 0..n {
        for j in 0..=w {
            if j + wv[i].0 <= w {
                chmax!(dp[i + 1][j + wv[i].0], dp[i][j] + wv[i].1);
            }
            chmax!(dp[i + 1][j], dp[i][j]);
        }
    }
    println!("{}", dp[n][w]);
}
